Update: Add ESLint API (refs eslint/rfcs#40) (#12939)
eslint/eslint: docs/developer-guide/nodejs-api.md

    * [defineParser()](#linter-defineparser)
    * [version](#linter-version)
* [linter (deprecated)](#linter-1)
* [CLIEngine (deprecated)](#cliengine)
* [RuleTester](#ruletester)
    * [Customizing RuleTester](#customizing-ruletester)
* [Deprecated APIs](#deprecated-apis)

---

## ESLint class

> The `ESLint` class is the facade class to use ESLint from Node.js application.

Reviewer: ```suggestion
The `ESLint` class is primary class to use in Node.js applications.
```

> This class pretty depends on the file system, so you cannot use it on browsers. If you want to lint code on browsers, use the [Linter](#linter) class instead.

Reviewer: ```suggestion
This class depends on the Node.js `fs` module and the file system, so you cannot use it in browsers. If you want to lint code on browsers, use the [Linter](#linter) class instead.
```

> For a simple example:

Reviewer: ```suggestion
Here's a simple example of using the `ESLint` class:
```

```js
const { ESLint } = require("eslint");

(async function main() {
  // 1. Create an instance.
  const eslint = new ESLint();

  // 2. Lint files.
  const results = await eslint.lintFiles(["lib/**/*.js"]);

  // 3. Format the results.
  const formatter = await eslint.loadFormatter("stylish");
  const resultText = formatter.format(results);

  // 4. Output it.
  console.log(resultText);
})().catch((error) => {
  process.exitCode = 1;
  console.error(error);
});
```

> For a simple example with autofix:

Reviewer: ```suggestion
And here is an example that autofixes lint problems:
```

* `options.rulePaths` (`string[]`)<br>
  Default is `[]`. An array of paths to directories to load custom rules from.
* `options.useEslintrc` (`boolean`)<br>
  Default is `true`. If `false` is present, ESLint doesn't load configuration files (`.eslintrc.*` files). Only the configuration of the constructor options is valid.

##### Autofix

* `options.fix` (`boolean | (message: LintMessage) => boolean`)<br>
  Default is `false`. If `true` is present, the [`eslint.lintFiles()`][eslint-lintfiles] and [`eslint.lintText()`][eslint-linttext] methods work in autofix mode. If a function is present, the methods pass each lint message to the function, then use only the lint messages that the function returned `true`.
* `options.fixTypes` (`("problem" | "suggestion" | "layout")[] | null`)<br>
  Default is `null`. The types of the rules that the [`eslint.lintFiles()`][eslint-lintfiles] and [`eslint.lintText()`][eslint-linttext] methods use for autofix.

> ##### Misc

Reviewer: ```suggestion
##### Cache-related
```

* `options.cache` (`boolean`)<br>
  Default is `false`. If `true` is present, the [`eslint.lintFiles()`][eslint-lintfiles] method caches lint results and uses it if each target file is not changed. Please mind that ESLint doesn't clear the cache when you upgrade ESLint plugins. In that case, you have to remove the cache file manually. The [`eslint.lintText()`][eslint-linttext] method doesn't use caches even if you pass the `options.filePath` to the method.
* `options.cacheLocation` (`string`)<br>
  Default is `.eslintcache`. The [`eslint.lintFiles()`][eslint-lintfiles] method writes caches into this file.

### ◆ eslint.lintFiles(patterns)

```js
const results = await eslint.lintFiles(patterns);
```

> This method checks the given files then returns the results.

Reviewer: ```suggestion
This method lints the files that match the glob patterns and then returns the results.
```

#### Parameters

* `patterns` (`string | string[]`)<br>
  The lint target files. This can contain any of file paths, directory paths, and glob patterns.

#### Return Value

* (`Promise<LintResult[]>`)<br>
  The promise that will be fulfilled with an array of [LintResult] objects.

### ◆ eslint.lintText(code, options)

```js
const results = await eslint.lintText(code, options);
```

> This method checks the given source code text then returns the results.

Reviewer: ```suggestion
This method lints the given source code text and then returns the results.
```

> If you want to use suitable configuration files to lint the text, passing the `options.filePath` option is important. This method loads configuration files with the same manner as the [`eslint.lintFiles()`][eslint-lintfiles] method checks the file the `options.filePath` option pointers. If the `options.filePath` option is not present, this method uses the configuration file at the current working directory (the `cwd` constructor option).

Reviewer: ```suggestion
If you want to use configuration files while linting the text, pass the `options.filePath` option. This method loads configuration files in the same manner as the [`eslint.lintFiles()`][eslint-lintfiles] method. If the `options.filePath` option is not present, this method uses the configuration file at the current working directory (the `cwd` constructor option).
```

Reviewer 2: > If you want to use configuration files while linting the text, pass the `options.filePath` option.

Sounds like it doesn't use configuration files if the `options.filePath` option is not present. But actually it uses configuration files on the cwd in that case.

> If the given `options.filePath` option value is an ignored path by your configurations, this method returns an empty array. If the `options.warnIgnored` option is set along with the `options.filePath` option, this method returns a [LintResult] object at least. In that case, the result may contain a warning that indicates the file was ignored.

Reviewer: ```suggestion
If the `options.filePath` value is configured to be ignored, this method returns an empty array. If the `options.warnIgnored` option is set along with the `options.filePath` option, this method returns a [LintResult] object. In that case, the result may contain a warning that indicates the file was ignored.
```

* It adds the `processor` setting if a legacy file extension processor is matched.
* It doesn't interpret the `env` setting to the `globals` and `parserOotions` settings, so the result object contains the `env` setting as is.

You can use the calculated configuration object for the debugging purpose.

#### Parameters

* `filePath` (`string`)<br>
  The path to the file you want to check the actual configuration. Directory paths are forbidden because ESLint cannot handle the `overrides` setting.

#### Return Value

* (`Promise<Object>`)<br>
  The promise that will be fulfilled with a configuration object.

### ◆ eslint.isPathIgnored(filePath)

```js
const isPathIgnored = await eslint.isPathIgnored(filePath);
```

This method checks if a given file is ignored by your configurations.

#### Parameters

* `filePath` (`string`)<br>
  The path to the file you want to check.

#### Return Value

* (`Promise<boolean>`)<br>
  The promise that will be fulfilled with whether the file is ignored or not. If the file is ignored then `true`.

### ◆ eslint.loadFormatter(nameOrPath)

```js
const formatter = await eslint.loadFormatter(nameOrPath);
```

> This method loads a formatter.

Reviewer 2: Hmm. I'm not sure what is the proper page, so I added more descriptions.

        * `"@foo"` will load `@foo/eslint-formatter`.
        * `"@foo/bar"` will load `@foo/eslint-formatter-bar`.
    * A path to the file that defines a formatter. The path must contain one or more path separators (`/`) in order to distinguish if it's a path or not.

#### Return Value

* (`Promise<Formatter>`)<br>
  The promise that will be fulfilled with a [Formatter] object.

### ◆ ESLint.version

```js
const version = ESLint.version;
```

The version string of ESLint. E.g. `"7.0.0"`.

This is not instance property.

### ◆ ESLint.outputFixes(results)

```js
await ESLint.outputFixes(results);
```

This method writes the modified code by the autofix feature into each file. If any modified files don't exist, this method does nothing.

This is not instance method.

#### Parameters

* `results` (`LintResult[]`)<br>
  The [LintResult] objects to write.

#### Return Value

* (`Promise<void>`)<br>
  The promise that will be fulfilled after all files are written.

### ◆ ESLint.getErrorResults(results)

```js
const filteredResults = ESLint.getErrorResults(results);
```

This method copies the given results and removes warnings. The returned value contains only errors.

> This is not instance method.

Author: ```suggestion
This is not an instance method.
```
or 

```suggestion
This is a static method.
```

#### Parameters

* `results` (`LintResult[]`)<br>
  The [LintResult] objects to filter.

#### Return Value

* (`LintResult[]`)<br>
  The filtered [LintResult] objects.

### ◆ LintResult type

> The `LintResult` value is the information of the linting result of each file. The [`eslint.lintFiles()`][eslint-lintfiles] and [`eslint.lintText()`][eslint-linttext] methods return it. It has following properties:

Author: ```suggestion
The `LintResult` value is the information of the linting result of each file. The [`eslint.lintFiles()`][eslint-lintfiles] and [`eslint.lintText()`][eslint-linttext] methods return it. It has the following properties:
```

* `filePath` (`string`)<br>

>   The absolute path to the file of this result. This may be the string `"<text>"` rather than a file path if the [`eslint.lintText()`][eslint-linttext] method returned it.

Author: Can we add an explanation of what `"<text>"` means (i.e. that it's a placeholder)? Or maybe link to some existing documentation if we already have some?

Reviewer 2: I added a description.

* `messages` (`LintMessage[]`)<br>
  The array of [LintMessage] objects.
* `fixableErrorCount` (`number`)<br>
  The number of errors that can be fixed automatically by the `fix` constructor option.
* `fixableWarningCount` (`number`)<br>
  The number of warnings that can be fixed automatically by the `fix` constructor option.
* `errorCount` (`number`)<br>
  The number of errors. This includes fixable errors.
* `warningCount` (`number`)<br>
  The number of warnings. This includes fixable warnings.
* `output` (`string | undefined`)<br>
  The modified source code text. This property is undefined if any fixable messages didn't exist.
* `source` (`string | undefined`)<br>
  The original source code text. This property is undefined if any messages didn't exist or the `output` property exists.
* `usedDeprecatedRules` (`{ ruleId: string; replacedBy: string[] }[]`)<br>
  The information about the deprecated rules that were used to check this file.

### ◆ LintMessage type

> The `LintMessage` value is the information of each linting error. The `messages` property of the [LintResult] type contains it. It has following properties:

Author: ```suggestion
The `LintMessage` value is the information of each linting error. The `messages` property of the [LintResult] type contains it. It has the following properties:
```
